server/src/test/java/org/elasticsearch/indices/cluster/ClusterStateChanges.java

```diff
@@ -277,9 +277,12 @@ public ClusterState applyFailedShards(ClusterState clusterState, List<FailedShar
     }
 
     public ClusterState applyStartedShards(ClusterState clusterState, List<ShardRouting> startedShards) {
-        List<StartedShardEntry> entries = startedShards.stream().map(startedShard ->
-            new StartedShardEntry(startedShard.shardId(), startedShard.allocationId().getId(), "shard started"))
-            .collect(Collectors.toList());
+        List<StartedShardEntry> entries = startedShards.stream()
+            .map(startedShard -> {
+                final IndexMetaData indexMetaData = clusterState.metaData().index(startedShard.shardId().getIndex());
+                final long primaryTerm = indexMetaData != null ? indexMetaData.primaryTerm(startedShard.shardId().id()) : 0L;
```

Comment:
this is simulating that we're always using the correct primary term, which is incorrect. For `IndicesClusterStateServiceRandomUpdateTests#randomlyUpdateClusterState` we could use the actual primary term of the shard, or record the primary term / shard routing at shard creation time like we do in `IndicesClusterStateService`. Perhaps an assertion could be that primary term is not updated on a shard routing when it is initializing.

Reply:
Good point